### PR TITLE
added UHD support for client and BS

### PR DESCRIPTION
**Pull request by tingjunchen**
- Added UHD support for the client using `kUseUHD` flag that can be passed to `cmake`
- Updated client radio config based on the `if (kUseUHD)` condition and the UHD-specific hardware settings. In particular, added function `RadioTXRX::loop_tx_rx_usrp_sync()` for the client to send pilot from the host
- Updated BA radio config for Agora and the UHD-specific hardware settings. In particular, added function `PacketTXRX::loop_tx_rx_usrp()` for the BS to (i) send beacon from the host, (ii) only receive relevant symbols on the uplink (i.e., P and U for now)
- Note: some functions in `txrx_usrp.cpp` are not needed but I left a copy of them from `txrx_argos.cpp` in case they could be of some use in the future
- Updated `CMakeLists.txt` to use `txrx_usrp` for USRP devices
- Added sample json files for USRP devices
- Cleaned up warnings during build

- Tested with a 2-antenna BS and a single-antenna client using two USRP x310s, with `ENABLE_MAC=off`
- Tested with a 2-antenna BS and two single-antenna clients using two USRP x310s, with `ENABLE_MAC=off`
- Note: need to comment out L257-263 and L282-283 in `src/agora/agora.cpp` for testing txrx streaming with UHD (these changes were not committed to avoid messing up the main agora file) -> fixed. Now Agora supports 2-antenna BS and we no longer need to comment out these lines

- (Update 10/04/2020) Merged USRP client TxRx with Argos in the sync case

### CODE REVIEW

File: src/agora/txrx/txrx_usrp.cpp [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+                rx_frame_start[frame_id % kNumStatsFrames] = rdtsc();
+                prev_frame_id = frame_id;
+            }
+        }
+    }
+    return 0;
+}
+
+struct Packet* PacketTXRX::recv_enqueue_usrp(
+    int tid, int radio_id, int rx_offset)
```
Comment (jianding17) [inline src/agora/txrx/txrx_usrp.cpp:111]: Do you need to keep this function? It seems only this one is used.
```
struct Packet* recv_enqueue_usrp(
     int tid, int radio_id, int rx_offset, int frame_id, int symbol_id);
```

File: src/agora/txrx/txrx_usrp.cpp [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        return NULL;
+    }
+
+    int frame_id = (int)(frameTime >> 32);
+    int symbol_id = (int)((frameTime >> 16) & 0xFFFF);
+    int ant_id = radio_id * nChannels;
+    for (int ch = 0; ch < nChannels; ++ch) {
+        new (pkt[ch]) Packet(frame_id, symbol_id, 0 /* cell_id */, ant_id + ch);
+        // move ptr & set status to full
+        // rx_buffer_status[rx_offset + ch] = 1; // has data, after it is read, it is set to 0
```
Comment (jianding17) [inline src/agora/txrx/txrx_usrp.cpp:146]: It should be safe to uncomment this line now. BTW, clang-format does not do line break for comments. We need to do it manually when the comment length is too long.

File: src/agora/txrx/txrx_usrp.cpp (hunk 1 above)
Comment (tingjunchen) [reply src/agora/txrx/txrx_usrp.cpp]: removed!

File: src/agora/txrx/txrx_usrp.cpp (hunk 2 above)
Comment (tingjunchen) [reply src/agora/txrx/txrx_usrp.cpp]: done.

File: src/client/txrx_client.cpp [hunk 3]
```diff
@@ -503,14 +500,21 @@ void* RadioTXRX::loop_tx_rx_argos_sync(int tid)
...
     size_t cursor(0);
     std::stringstream sout;
+
+    // Keep receiving one frame of data until a beacon is found
+    // Perform initial beacon detection every 10 frames
     while (c->running && sync_index < 0) {
-        int r = radio->radioRx(
-            radio_id, frm_rx_buff.data(), frm_num_samps, rxTime);
+        int r;
+        for (int find_beacon_retry = 0; find_beacon_retry < 10;
```
Comment (jianding17) [inline src/client/txrx_client.cpp:508]: It is better to replace `10` here with a global variable. You can add it to src/common/Symbols.hpp, e.g., `static constexpr size_t kBeaconDetectInterval = 10`, and also add the corresponding comment there.

File: src/client/txrx_client.cpp [hunk 4]
```diff
@@ -285,12 +278,14 @@ int RadioTXRX::dequeue_send_argos(int tid, long long time0)
     long long txTime(0);
     int r;
 
-    if (!c->hw_framer) {
+    // transmit pilot
+    if (kUseUHD || !c->hw_framer) {
```
[the remark this answers is not in the document]
Comment (tingjunchen) [reply src/client/txrx_client.cpp:282]: fixed!

File: src/client/txrx_client.cpp [hunk 5]
```diff
@@ -272,6 +262,9 @@ int RadioTXRX::dequeue_send_argos(int tid, long long time0)
     int num_samps = c->sampsPerSymbol;
     int frm_num_samps = num_samps * c->symbol_num_perframe;
 
+    // for UHD devices, first pilot should not be with the END_BURST flag
+    int flagsTxPilot = (kUseUHD && c->nChannels == 2) ? 1 : 2;
```
[the remark this answers is not in the document]
Comment (tingjunchen) [reply src/client/txrx_client.cpp:266]: fixed

File: src/client/txrx_client.cpp [hunk 6]
```diff
@@ -272,6 +262,9 @@ int RadioTXRX::dequeue_send_argos(int tid, long long time0)
     int num_samps = c->sampsPerSymbol;
     int frm_num_samps = num_samps * c->symbol_num_perframe;
 
+    // for UHD devices, first pilot should not be with the END_BURST flag
```
[the remark this answers is not in the document]
Comment (tingjunchen) [reply src/client/txrx_client.cpp:265]: done

File: src/client/txrx_client.cpp (hunk 3 above)
Comment (tingjunchen) [reply src/client/txrx_client.cpp]: fixed!

**Pull request merged**